language/classify_text/classify_text_tutorial.py

```diff
@@ -0,0 +1,251 @@
+        result[category.name] = category.confidence
+
+    if verbose:
+        print(text)
+        for category in categories:
+            print(u'=' * 20)
+            print(u'{:<16}: {}'.format('category', category.name))
+            print(u'{:<16}: {}'.format('confidence', category.confidence))
+
+    return result
+# [END def_classify]
+
+
+# [START def_index]
+def index(path, index_file):
+    """Classify each text file in the directory and write
+    the results to the index_file.
+    """
+
+    result = {}
+    for filename in os.listdir(path):
+        file_path = os.path.join(path, filename)
+
+        if not os.path.isfile(file_path):
+            continue
+
+        try:
+            with open(file_path, 'r') as f:
+                text = f.read()
+                categories = classify(text, verbose=False)
+
+                result[filename] = categories
+        except:
+            print('Failed to process {}'.format(file_path))
+
+    with open(index_file, 'w') as f:
+        json.dump(result, f)
+
+    print('Texts indexed in file: {}'.format(index_file))
+    return result
+# [END def_index]
+
+
+# [START def_split_labels]
+def split_labels(categories):
+    """The category labels are of the form "/a/b/c" up to three levels,
+    for example "/Computers & Electronics/Software", and these labels
+    are used as keys in the categories dictionary, whose values are
+    confidence scores.
+
+    The split_labels function splits the keys into individual levels
+    while duplicating the confidence score, which allows a natural
+    boost in how we calculate similarity when more levels are in common.
+
+    Example:
+    If we have
+
+    x = {"/a/b/c": 0.5}
+    y = {"/a/b": 0.5}
+    z = {"/a": 0.5}
+
+    Then x and y are considered more similar than y and z.
+    """
+    _categories = {}
+    for name, confidence in categories.iteritems():
+        labels = [label for label in name.split('/') if label]
+        for label in labels:
+            _categories[label] = confidence
+
+    return _categories
+# [END def_split_labels]
+
+
+# [START def_similarity]
+def similarity(categories1, categories2):
+    """Cosine similarity of the categories treated as sparse vectors."""
+    categories1 = split_labels(categories1)
+    categories2 = split_labels(categories2)
+
+    norm1 = np.linalg.norm(categories1.values())
+    norm2 = np.linalg.norm(categories2.values())
+
+    # Return the smallest possible similarity if either categories is empty.
+    if norm1 == 0 or norm2 == 0:
+        return 0.0
+
+    # Compute the cosine similarity.
+    dot = 0.0
+    for label, confidence in categories1.iteritems():
+        dot += confidence * categories2.get(label, 0.0)
+
+    return dot / (norm1 * norm2)
+# [END def_similarity]
+
+
+# [START def_query]
+def query(index_file, text, n_top=3):
+    """Find the indexed files that are the most similar to
+    the query text.
+    """
+
+    with open(index_file, 'r') as f:
```

Comment:
Throughout this file, please use `io.open` over `open`.

Reply:
done.